backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/utils/VdsCpuUnitPinningHelper.java

```diff
@@ -129,64 +110,47 @@ private void previewPinOfPendingExclusiveCpus(List<VdsCpuUnit> cpuTopology, Map<
-        for (int core : getOnlineCores(cpuTopology, socket)) {
-            List<VdsCpuUnit> sharedCpus = getFreeCpusInCore(cpuTopology, socket, core);
-            int sharedCpusInCore = sharedCpus.size();
-            List<Integer> sharedCpuIds = sharedCpus.stream().map(VdsCpuUnit::getCpu).collect(Collectors.toCollection(LinkedList::new));
-            Iterator<Integer> sharedCpusIterator = sharedCpuIds.iterator();
-            if (cpusLeft == 0) {
-                return cpusLeft;
-            }
-            if (wholeCore) {
-                if (cpusLeft >= hostThreadsPerCore && sharedCpusInCore == hostThreadsPerCore) {
-                    // take the whole core
-                    while (sharedCpusInCore-- > 0) {
-                        VdsCpuUnit cpuToTake = getCpu(sharedCpus, socket, core, sharedCpusIterator.next());
-                        cpusToBeAllocated.add(cpuToTake);
-                        cpuToTake.pinVm(vmId, cpuPinningPolicy);
-                        cpusLeft--;
+    private void allocateCpus(List<VdsCpuUnit> cpuTopology, VM vm, List<VdsCpuUnit> cpusToBeAllocated) {
+        int socketsLeft = vm.getNumOfSockets();
+        int onlineSockets = getOnlineSockets(cpuTopology).size();
+        while (onlineSockets > 0 && socketsLeft > 0) {
+            List<VdsCpuUnit> cpusInChosenSocket = getMaxFreedSocket(cpuTopology);
+            // coreCount is based on the VM topology
+            int coreCount = 0;
+            for (int core : getOnlineCores(cpusInChosenSocket)) {
+                List<VdsCpuUnit> freeCpusInCore = getFreeCpusInCore(cpusInChosenSocket, core);
+                int coreThreads = freeCpusInCore.size();
+                while (coreThreads >= vm.getThreadsPerCpu()) {
+                    for (int thread = 0; thread < vm.getThreadsPerCpu(); thread++) {
+                        VdsCpuUnit cpuUnit = freeCpusInCore.remove(0);
+                        cpuUnit.pinVm(vm.getId(), vm.getCpuPinningPolicy());
+                        cpusToBeAllocated.add(cpuUnit);
                     }
+                    coreCount++;
+                    coreThreads -= vm.getThreadsPerCpu();
                 }
-            } else {
-                while (sharedCpusInCore-- > 0 && cpusLeft > 0) {
-                    VdsCpuUnit cpuToTake = getCpu(sharedCpus, socket, core, sharedCpusIterator.next());
-                    cpusToBeAllocated.add(cpuToTake);
-                    cpuToTake.pinVm(vmId, cpuPinningPolicy);
-                    cpusLeft--;
+            }
+            while (socketsLeft > 0 && coreCount >= vm.getCpuPerSocket()) {
+                // If we manage to allocate enough cores to say it is enough for virtual socket
+                // drop one socket. It is possible that within one physical socket we can allocate multiple virtual
+                // sockets, if so - take it as it's better.
+                socketsLeft--;
+                coreCount -= vm.getCpuPerSocket();
+            }
```

Comment:
I see, so this is the same logic as above but this time after allocating the cpus - so can we do something like socketsLeft -= coreCount / (cores-per-socket * threads-per-core) ?

Reply:
After removing the below section and adding another condition to the loops above, yes. but without the threads (they are suppose to be included already).